src/engine/sparqlExpressions/PrefilterExpressionIndex.cpp

```diff
@@ -134,39 +134,100 @@ static auto getSetUnion(const std::vector<BlockMetadata>& blocks1,
+}
+
+//______________________________________________________________________________
+// Similar to `getRelevantBlocksFromIdRanges`, but this function retrieves the
+// relevant (and save) complementing `BlockMetadata` values w.r.t. `input` and
+// `relevantIdRanges`.
+static std::vector<BlockMetadata> getComplementBlocksForRelevantIdRanges(
+    const std::vector<RandomItPair>& relevantIdRanges,
+    std::span<const BlockMetadata> input, const RandomIt& valueIdsInputBegin,
+    const RandomIt& valueIdsInputEnd) {
+  // If relevantIdRanges is empty, the complement are simply all `BlockMetadata`
+  // values (input).
+  if (relevantIdRanges.empty()) {
+    return std::vector<BlockMetadata>{input.begin(), input.end()};
+  }
+  // Use helper getInputAdjustedRangesIndices (w.r.t. std::span<const
+  // BlockMetadata> input) and bind relevant constant arguments.
+  const auto getInputAdjustedRange = std::bind(
+      getInputAdjustedRangeIndices, input.begin(), input.end(),
+      valueIdsInputBegin, std::placeholders::_1, std::placeholders::_2);
+  // Section retrieve complementing BlockMetadata values.
+  std::vector<BlockMetadata> complementBlocks;
+  complementBlocks.reserve(input.size());
+  // We implicitly handle the first complement range:
+  // input[input.begin(), (first) inputAdjustedBegin]
+  RandomIt currentBegin = valueIdsInputBegin;
+  for (const auto& [firstRelevantIt, lastRelevantIt] : relevantIdRanges) {
+    const auto& [inputAdjustedBegin, inputAdjustedEnd] =
+        getInputAdjustedRange(currentBegin, firstRelevantIt);
+    complementBlocks.insert(complementBlocks.end(), inputAdjustedBegin,
+                            inputAdjustedEnd);
+    currentBegin = lastRelevantIt;
+  }
+  // Handle the complementing range:
+  // input[(last) inputAdjustedBegin, input.end()]
+  if (currentBegin < valueIdsInputEnd) {
+    const auto& [inputAdjustedBegin, inputAdjustedEnd] =
+        getInputAdjustedRange(currentBegin, valueIdsInputEnd);
+    complementBlocks.insert(complementBlocks.end(), inputAdjustedBegin,
+                            inputAdjustedEnd);
+  }
+  complementBlocks.shrink_to_fit();
+  return complementBlocks;
+}
+
 //______________________________________________________________________________
 static std::vector<BlockMetadata> getRelevantBlocksFromIdRanges(
     const std::vector<RandomItPair>& relevantIdRanges,
-    std::span<const BlockMetadata> input, std::vector<ValueId>& valueIdsInput) {
+    std::span<const BlockMetadata> input, const RandomIt& valueIdsInputBegin) {
   // The vector for relevant BlockMetadata values which contain ValueIds
   // defined as relevant by relevantIdRanges.
   std::vector<BlockMetadata> relevantBlocks;
   // Reserve memory, input.size() is upper bound.
   relevantBlocks.reserve(input.size());
 
+  const auto getInputAdjustedRange = std::bind(
+      getInputAdjustedRangeIndices, input.begin(), input.end(),
+      valueIdsInputBegin, std::placeholders::_1, std::placeholders::_2);
```

Comment:
std::bind is somewhat outdated, either use an explicit lambda, or (in this case) use `std::bind_front`.